File: perl/ray.pl
#!/usr/bin/perl
use strict;
use warnings;

our $VERSION=q(1.0.0);

my $INFINITY=1.0e+300;
my $EPSILON=0.00000000000000001;
my $DELTA=sqrt($EPSILON);

my $ZEROVEC={X=>0.0,Y=>0.0,Z=>0.0};

my $OVERSAMPLESIZE=2;
my $RESOLUTION=256;

my $ST_GROUP=0;
my $ST_SPHERE=1;

################################################################
sub multiply_vec{
    my($s, $r)=@_;
    return ({ X=>$s*$r->{X}, Y=>$s*$r->{Y}, Z=>$s*$r->{Z} });
}

sub vec_add{
    my($A, $B)=@_;
    return ({ X=>($A->{X}+$B->{X}), Y=>($A->{Y}+$B->{Y}), Z=>($A->{Z}+$B->{Z}) });
}

sub vec_subtract{
    my($A, $B)=@_;
    return ({ X=>($A->{X}-$B->{X}), Y=>($A->{Y}-$B->{Y}), Z=>($A->{Z}-$B->{Z}) });
}

sub vec_dot{
    my($A, $B)=@_;
    return ( ($A->{X} * $B->{X}) + ($A->{Y} * $B->{Y}) + ($A->{Z} * $B->{Z}) )
};

sub vec_unitise{
    my ($R)=@_;
    return( multiply_vec((1.0 / sqrt(vec_dot($R,$R))), $R) );
}

sub ray_sphere{
    my ($Orig, $Dir, $Center, $Radius)=@_;
    my $V=vec_subtract($Center,$Orig);
    my $B=vec_dot($V,$Dir);
    my $D= (($B * $B) - vec_dot($V,$V) + ($Radius * $Radius));
    if ($D<0.0) {
	return($INFINITY);
    }else{
	my$D2=sqrt($D);
	my $T2=$B + $D2;
	if ($T2 < 0.0) {
	    return($INFINITY);
	}else{
	    my $T1=$B - $D2;
	    if ($T1 > 0.0){
		return($T1);
	    }else{
		return($T2);
	    }
	}
    }
}

sub ofscene{
    my($Orig,$Dir,  $LN, $Scene)=@_;

    if ($Scene->{TYPE} == $ST_SPHERE){
	my $Lp=ray_sphere($Orig, $Dir, $Scene->{CENTER},  $Scene->{RADIUS});
	if ($Lp >= $LN->{L}){
	    return ($LN);
	}else{
	    return ({L=>$Lp,N=>vec_unitise(vec_subtract(vec_add ($Orig, multiply_vec($Lp, $Dir)),$Scene->{CENTER}))});
	}
    }elsif ($Scene->{TYPE} == $ST_GROUP){
	my $Lp=ray_sphere($Orig, $Dir, $Scene->{CENTER},  $Scene->{RADIUS});
	if ($Lp>= $LN->{L}){
	    return ($LN);
	}else{
	    # lists:foldl(fun (S, LN) -> ofscene(Orig,Dir, S, LN) end, {L,N}, Scenes)

	    my $qLN=$LN;
	    foreach my $S (@{$Scene->{SCENES}}){$qLN=ofscene($Orig,$Dir, $qLN, $S);}
	    return ($qLN);
	}
    }
}

sub intersect{
    my ($Orig,$Dir, $Scene)=@_;
    return(ofscene($Orig,$Dir, {L=>$INFINITY, N=>$ZEROVEC}, $Scene));
}

sub ray_trace{
    my ($Light, $Orig, $Dir, $Scene)=@_;
    my $LN=intersect($Orig, $Dir, $Scene);
    if ($LN->{L} >= $INFINITY){
	return(0.0);
    }else{
	my $G=0.0-vec_dot($LN->{N}, $Light);
	if ($G<=0.0){
	    return(0.0);
	}else{
	    my $OrigP=vec_add($Orig, vec_add(multiply_vec($LN->{L}, $Dir),  multiply_vec($DELTA,$LN->{N})));
	    my $DirP=vec_subtract($ZEROVEC, $Light);
	    my $LqN=intersect($OrigP, $DirP, $Scene);
	    if ($LqN->{L} >= $INFINITY){
		return($G);
	    }else{
		return(0.0);
	    }
	}
    }
}

sub create{
    my ($Level, $R, $X, $Y, $Z)=@_;

    my $Obj={
	TYPE=>$ST_SPHERE,
	CENTER=>{X=>$X,Y=>$Y,Z=>$Z},
	RADIUS=>$R};
    if ($Level == 1){
	return($Obj);
    }else{
	my $Rp = 3.0 * $R / sqrt(12.0);
	return(
	    {TYPE=>$ST_GROUP,
	     CENTER=>{X=>$X,Y=>$Y,Z=>$Z},
	     RADIUS=>3.0 * $R,
	     SCENES=>[
		 $Obj,
		 create(($Level-1), (0.5*$R), $X-$Rp, $Y+$Rp, $Z-$Rp),
		 create(($Level-1), (0.5*$R), $X+$Rp, $Y+$Rp, $Z-$Rp),
		 create(($Level-1), (0.5*$R), $X-$Rp, $Y+$Rp, $Z+$Rp),
		 create(($Level-1), (0.5*$R), $X+$Rp, $Y+$Rp, $Z+$Rp)
		 ] }
	    );
    }
}
################################################################
sub pixel{
    my ($Eye, $X,$Y, $Scene)=@_;
			
    my $rsum=0;
    for my $Dx (0..$OVERSAMPLESIZE-1){
	for my $Dy (0..$OVERSAMPLESIZE-1){
	    $rsum+=ray_trace($Eye,
			     {X=>0.0, Y=>0.0, Z=>-4.0},
			     vec_unitise({ X=>$X+($Dx/$OVERSAMPLESIZE)-($RESOLUTION/2.0),
					   Y=>$RESOLUTION-1.0-$Y+($Dy/$OVERSAMPLESIZE)-($RESOLUTION/2.0),
					   Z=>$RESOLUTION}),
			     $Scene)
	}
    }
    return (int(0.5 + 255.0 * ($rsum)/($OVERSAMPLESIZE*$OVERSAMPLESIZE)));
}

################################################################
sub start{
    my $Level=5; # Number of Spheres
    my$Eye=vec_unitise({X=>-3.0,Y=>-3.0,Z=>2.0});
    my$Scene=create($Level,  1.0,  0.0,-1.0,0.0); # Scene tree
    trace($Eye, $Scene);
    return;
}

sub trace{
    my ($Eye, $Scene)=@_;

    printf(STDOUT "P2\n%s %s\n255\n",$RESOLUTION,$RESOLUTION);
    for my $Y (0..$RESOLUTION-1){
	printf(STDOUT "%s\n", join(q( ),
				   map {pixel($Eye,$_,$Y,$Scene)} 0..$RESOLUTION-1
	       )
	    );
    }
    return;
}

################################################################
start();

File: go/ray.go
package main

import (
	"fmt"
	"math"
	"os"
	"runtime"
)

const (
	OVERSAMPLESIZE = 4
	RESOLUTION     = 512
	LEVEL          = 6
)

type Vector struct {
	x float64
	y float64
	z float64
}

type Hit struct {
	l float64
	v Vector
}

type Object struct {
	kind   int
	center Vector
	radius float64
	objs   []Object
}

const (
	INFINITY = 1.0e+300
	EPSILON  = 0.00000000000000001

	ST_GROUP  = 0
	ST_SPHERE = 1
)

var ZEROVEC Vector = Vector{x: 0.0, y: 0.0, z: 0.0}
var DELTA float64 = math.Sqrt(EPSILON)

//----------------------------------------------------------------
func multiply_vec(s Vector, r float64) Vector {
	return Vector{x: r * s.x, y: r * s.y, z: r * s.z}
}

func vec_add(a Vector, b Vector) Vector       {
	return Vector{x: (a.x + b.x), y: (a.y + b.y), z: (a.z + b.z)}
}

func vec_subtract(a Vector, b Vector) Vector {
	return Vector{x: (a.x - b.x), y: (a.y - b.y), z: (a.z - b.z)}
}

func vec_dot(a Vector, b Vector) float64 {
	return (a.x * b.x) + (a.y * b.y) + (a.z * b.z)
}

func vec_unitise(a Vector) Vector {
	return multiply_vec(a, (float64(1.0) / math.Sqrt(vec_dot(a, a))))
}

//----------------------------------------------------------------
func ray_sphere(orig Vector, dir Vector, center Vector, radius float64) float64 {
	v := vec_subtract(center, orig)
	b := vec_dot(v, dir)
	d := ((b * b) - vec_dot(v, v) + (radius * radius))

	if d < 0.0 {
		return INFINITY
	} else {
		d2 := math.Sqrt(d)
		t2 := b + d2

		if t2 < 0.0 {
			return INFINITY
		} else {
			t1 := b - d2
			if t1 > 0.0 {
				return t1
			} else {
				return t2
			}
		}
	}
}

func ofscene(orig Vector, dir Vector, ln Hit, scene Object) Hit {
	if scene.kind == ST_SPHERE {
		lp := ray_sphere(orig, dir, scene.center, scene.radius)

		if lp >= ln.l {
			return ln
		}
		return Hit{l: lp, v: vec_unitise(vec_subtract(vec_add(orig, multiply_vec(dir, lp)), scene.center))}

	} else if scene.kind == ST_GROUP {
		lp := ray_sphere(orig, dir, scene.center, scene.radius)
		qln := ln

		if lp >= ln.l {
			return ln
		} else {
			for _, s := range scene.objs {
				qln = ofscene(orig, dir, qln, s)
			}
		}
		return qln
	}
	return ln // should never reach this
}

func intersect(orig Vector, dir Vector, scene Object) Hit {
	return ofscene(orig, dir, Hit{l: INFINITY, v: ZEROVEC}, scene)
}

func ray_trace(light Vector, orig Vector, dir Vector, scene Object) float64 {
	ln := intersect(orig, dir, scene)

	if ln.l >= INFINITY {
		return 0.0
	} else {
		g := 0.0 - vec_dot(ln.v, light)
		if g <= 0.0 {
			return 0.0
		} else {
			origp := vec_add(orig, vec_add(multiply_vec(dir, ln.l), multiply_vec(ln.v, DELTA)))
			dirp := vec_subtract(ZEROVEC, light)
			lqn := intersect(origp, dirp, scene)

			if lqn.l >= INFINITY {
				return g
			}

			return 0.0
		}
	}
}

func create(level int, r float64, x float64, y float64, z float64) Object {
	obj := Object{
		kind:   ST_SPHERE,
		center: Vector{x: x, y: y, z: z},
		radius: r,
		objs:   nil,
	}

	if level == 1 {
		return obj
	} else {
		rp := 3.0 * r / math.Sqrt(12.0)
		return Object{
			kind:   ST_GROUP,
			center: Vector{x: x, y: y, z: z},
			radius: 3.0 * r,
			objs: []Object{
				obj,
				create((level - 1), (0.5 * r), x-rp, y+rp, z-rp),
				create((level - 1), (0.5 * r), x+rp, y+rp, z-rp),
				create((level - 1), (0.5 * r), x-rp, y+rp, z+rp),
				create((level - 1), (0.5 * r), x+rp, y+rp, z+rp),
			},
		}
	}
}

//----------------------------------------------------------------
func pixel(eye Vector, x int, y int, scene Object) int {

	var rsum float64 = float64(0.0)

	for dx := 0; dx <= (OVERSAMPLESIZE - 1); dx++ {
		for dy := 0; dy <= (OVERSAMPLESIZE - 1); dy++ {
			rsum = rsum + ray_trace(
				eye,
				Vector{x: 0.0, y: 0.0, z: -4.0},
				vec_unitise(
					Vector{
						x: float64(x) + (float64(dx) / float64(OVERSAMPLESIZE)) - (float64(RESOLUTION) / 2.0),
						y: float64(RESOLUTION) - 1.0 - float64(y) + (float64(dy) / float64(OVERSAMPLESIZE)) - (float64(RESOLUTION) / 2.0),
						z: float64(RESOLUTION),
					}),
				scene)
		}
	}
	return int(math.Floor(float64(0.5) + (float64(255.0) * ((rsum) / float64(OVERSAMPLESIZE*OVERSAMPLESIZE)))))
}

//----------------------------------------------------------------
func trace_single(eye Vector, scene Object) {
	fmt.Printf("P2\n%d %d\n255\n", RESOLUTION, RESOLUTION)
	for y := 0; y <= (RESOLUTION - 1); y++ {
		for x := 0; x <= (RESOLUTION - 1); x++ {
			if x > 0 {
				fmt.Printf(" ")
			}
			fmt.Printf("%d", pixel(eye, x, y, scene))
		}
		fmt.Printf("\n")
	}
}

//................................................................
func trace_goroutine (eye Vector, scene Object) {

	type workerdata struct {
		y int
		d [RESOLUTION]int
	}

	var numWorkers int = runtime.NumCPU() * 2
	runtime.GOMAXPROCS(numWorkers)

	image:= make([][RESOLUTION]int,RESOLUTION)

	done := make(chan struct{}, RESOLUTION)
	work := make(chan int, RESOLUTION)
	result := make(chan workerdata, RESOLUTION)

	// setup workers
	for i := 0; i < numWorkers; i++ {
		go func(workerid int, d chan struct{}, w chan int, r chan workerdata) {
			fmt.Fprintf(os.Stderr, "%02d: started\n", workerid)

			var workresult workerdata;
			for y := range w {
//				fmt.Fprintf(os.Stderr, "%02d: working y=%d\n", workerid, y)
				workresult.y=y;
				for x := 0; x <= (RESOLUTION - 1); x++ {
					workresult.d[x]=pixel(eye, x, y, scene)
				}
				select {
				case r <- workresult:
//					fmt.Fprintf(os.Stderr, "%02d: sent y=%d\n", workerid, y)
				case <-d:
					return
				}
			}
		}(i, done,work,result)
	}

	// fill up work queue
	for y := 0; y <= (RESOLUTION - 1); y++ { work <- y }

	// reap results
	for y := 0; y <= (RESOLUTION - 1); y++ {
		r := <- result
//		fmt.Fprintf(os.Stderr, "XX: result y=%d\n", r.y)
		image[r.y] = r.d
	}
	
	// display results
	fmt.Printf("P2\n%d %d\n255\n", RESOLUTION, RESOLUTION)
	for y := 0; y <= (RESOLUTION - 1); y++ {
		for x := 0; x <= (RESOLUTION - 1); x++ {
			if x > 0 {
				fmt.Printf(" ")
			}
			fmt.Printf("%d", image[y][x])
		}
		fmt.Printf("\n")
	}
}

//----------------------------------------------------------------
func main() {
	level := LEVEL // Number of Spheres
	eye := vec_unitise(Vector{x: -3.0, y: -3.0, z: 2.0})
	scene := create(level, 1.0, 0.0, -1.0, 0.0) // Scene tree

//	trace_single(eye, scene)
	trace_goroutine(eye, scene)
}
